File: tests/python3/src/T006_MessageFieldIterator.py
#!/usr/bin/env python3

import libgmsec_python3 as lp
from Test import Test


class Test_MessageFieldIterator(Test):

    def test(self):
        self.test_all_fields()
        self.test_header_fields()
        self.test_non_header_fields()
        self.test_reset()


    def test_all_fields(self):
        lp.log_info("Test ALL_FIELDS")

        msg = TestMessage.instance().get_message()

        iter = msg.get_field_iterator( lp.MessageFieldIterator.Selector_ALL_FIELDS )

        fieldsFound = 0

        while iter.has_next():
            field = iter.next()

            fieldsFound += 1

            if fieldsFound == 1:
                self.check("Unexpected field type 1", field.get_type() == lp.Field.Type_STRING)
                self.check("Unexpected field name 1", field.get_name() == "FIELD-1")
                self.check("Unexpected field value 1", field.get_string_value() == "1")

            elif fieldsFound == 2:
                self.check("Unexpected field type 2", field.get_type() == lp.Field.Type_STRING)
                self.check("Unexpected field name 2", field.get_name() == "FIELD-2")
                self.check("Unexpected field value 2", field.get_string_value() == "2")

            else:
                self.check("Unexpected field type 3", field.get_type() == lp.Field.Type_I32)
                self.check("Unexpected field name 3", field.get_name() == "FIELD-3")
                self.check("Unexpected field value 3", field.get_integer_value() == 3)

        self.check("Unexpected field count", fieldsFound == 3)


    def test_header_fields(self):
        lp.log_info("Test HEADER_FIELDS")

        msg = TestMessage.instance().get_message()

        iter = msg.get_field_iterator( lp.MessageFieldIterator.Selector_HEADER_FIELDS )

        fieldsFound = 0

        while iter.has_next():
            field = iter.next()

            fieldsFound += 1

            if fieldsFound == 1:
                self.check("Unexpected field type 1", field.get_type() == lp.Field.Type_STRING)
                self.check("Unexpected field name 1", field.get_name() == "FIELD-1")
                self.check("Unexpected field value 1", field.get_string_value() == "1")

            elif fieldsFound == 2:
                self.check("Unexpected field type 2", field.get_type() == lp.Field.Type_STRING)
                self.check("Unexpected field name 2", field.get_name() == "FIELD-2")
                self.check("Unexpected field value 2", field.get_string_value() == "2")

            else:
                self.check("Unexpected field", False)

        self.check("Unexpected field count", fieldsFound == 2)


    def test_non_header_fields(self):
        lp.log_info("Test NON_HEADER_FIELDS")

        msg = TestMessage.instance().get_message()

        iter = msg.get_field_iterator( lp.MessageFieldIterator.Selector_NON_HEADER_FIELDS )

        fieldsFound = 0

        while iter.has_next():
            field = iter.next()

            fieldsFound += 1

            if fieldsFound == 1:
                self.check("Unexpected field type 3", field.get_type() == lp.Field.Type_I32)
                self.check("Unexpected field name 3", field.get_name() == "FIELD-3")
                self.check("Unexpected field value 3", field.get_integer_value() == 3)

            else:
                self.check("Unexpected field", False)

        self.check("Unexpected field count", fieldsFound == 1)


    def test_reset(self):
        lp.log_info("Test reset()")

        msg = TestMessage.instance().get_message()

        iter = msg.get_field_iterator()

        fieldsFound = 0

        while iter.has_next():
            iter.next()
            fieldsFound += 1

        self.check("Unexpected field count", fieldsFound == 3)

        self.check("Should no more field(s) to iterate over", iter.has_next() == False)

        iter.reset()

        self.check("Should have field(s) to iterator over", iter.has_next())

        fieldsFound = 0

        while iter.has_next():
            iter.next()
            fieldsFound += 1

        self.check("Unexpected field count", fieldsFound == 3)


class TestMessage:

    s_instance = None

    @staticmethod
    def instance():
        if TestMessage.s_instance == None:
            TestMessage()
        return TestMessage.s_instance


    def __init__(self):
        TestMessage.s_instance = self
        self.m_msg = lp.Message()

        field1 = lp.StringField("FIELD-1", "1", True)
        field2 = lp.StringField("FIELD-2", "2", True)
        field3 = lp.I32Field("FIELD-3", 3)

        self.m_msg.add_field( field1 )
        self.m_msg.add_field( field2 )
        self.m_msg.add_field( field3 )


    def get_message(self):
        return self.m_msg


Test.run('MessageFieldIterator', Test_MessageFieldIterator())

File: tests/perl/src/T006_MessageFieldIterator.pl
#!/usr/bin/env perl

use strict;
use Test;
use libgmsec_perl;


sub Test_MessageFieldIterator
{
	my ($test) = @_;

	$test->setDescription("MessageFieldIterator");

	test_all_fields($test);
	test_header_fields($test);
	test_non_header_fields($test);
	test_reset($test);
}


sub test_all_fields
{
	libgmsec_perl::logInfo("Test ALL_FIELDS");

	my ($test) = @_;

	my $msg = libgmsec_perl::Message->new();
    my $field1 = libgmsec_perl::StringField->new("FIELD-1", "1", 1);
    my $field2 = libgmsec_perl::StringField->new("FIELD-2", "2", 1);
    my $field3 = libgmsec_perl::I32Field->new("FIELD-3", 3);

	$msg->addField( $field1 );
	$msg->addField( $field2 );
	$msg->addField( $field3 );

	my $iter = $msg->getFieldIterator( $libgmsec_perl::MessageFieldIterator::Selector_ALL_FIELDS );

	my $fieldsFound = 0;

	while ($iter->hasNext())
	{
		my $field = $iter->next();

		$fieldsFound += 1;

		if ($fieldsFound == 1)
		{
			$test->check("Unexpected field type 1", $field->getType() == $libgmsec_perl::Field::Type_STRING);
			$test->check("Unexpected field name 1", $field->getName() eq "FIELD-1");
			$test->check("Unexpected field value 1", $field->getStringValue() eq "1");
		}
		elsif ($fieldsFound == 2)
		{
			$test->check("Unexpected field type 2", $field->getType() == $libgmsec_perl::Field::Type_STRING);
			$test->check("Unexpected field name 2", $field->getName() eq "FIELD-2");
			$test->check("Unexpected field value 2", $field->getStringValue() eq "2");
		}
		else
		{
			$test->check("Unexpected field type 3", $field->getType() == $libgmsec_perl::Field::Type_I32);
			$test->check("Unexpected field name 3", $field->getName() eq "FIELD-3");
			$test->check("Unexpected field value 3", $field->getIntegerValue() == 3);
		}
	}

	$test->check("Unexpected field count", $fieldsFound == 3);
}


sub test_header_fields
{
	libgmsec_perl::logInfo("Test HEADER_FIELDS");

	my ($test) = @_;

	my $msg = libgmsec_perl::Message->new();
    my $field1 = libgmsec_perl::StringField->new("FIELD-1", "1", 1);
    my $field2 = libgmsec_perl::StringField->new("FIELD-2", "2", 1);
    my $field3 = libgmsec_perl::I32Field->new("FIELD-3", 3);

	$msg->addField( $field1 );
	$msg->addField( $field2 );
	$msg->addField( $field3 );

	my $iter = $msg->getFieldIterator( $libgmsec_perl::MessageFieldIterator::Selector_HEADER_FIELDS );

	my $fieldsFound = 0;

	while ($iter->hasNext())
	{
		my $field = $iter->next();

		$fieldsFound += 1;

		if ($fieldsFound == 1)
		{
			$test->check("Unexpected field type 1", $field->getType() == $libgmsec_perl::Field::Type_STRING);
			$test->check("Unexpected field name 1", $field->getName() eq "FIELD-1");
			$test->check("Unexpected field value 1", $field->getStringValue() eq "1");
		}
		elsif ($fieldsFound == 2)
		{
			$test->check("Unexpected field type 2", $field->getType() == $libgmsec_perl::Field::Type_STRING);
			$test->check("Unexpected field name 2", $field->getName() eq "FIELD-2");
			$test->check("Unexpected field value 2", $field->getStringValue() eq "2");
		}
		else
		{
            libgmsec_perl::logInfo("Unexpected field: " . $field->getName());
			$test->check("Unexpected field", 0);
		}
	}

	$test->check("Unexpected field count", $fieldsFound == 2);
}


sub test_non_header_fields
{
	libgmsec_perl::logInfo("Test NON_HEADER_FIELDS");

	my ($test) = @_;

	my $msg = libgmsec_perl::Message->new();
    my $field1 = libgmsec_perl::StringField->new("FIELD-1", "1", 1);
    my $field2 = libgmsec_perl::StringField->new("FIELD-2", "2", 1);
    my $field3 = libgmsec_perl::I32Field->new("FIELD-3", 3);

	$msg->addField( $field1 );
	$msg->addField( $field2 );
	$msg->addField( $field3 );

	my $iter = $msg->getFieldIterator( $libgmsec_perl::MessageFieldIterator::Selector_NON_HEADER_FIELDS );

	my $fieldsFound = 0;

	while ($iter->hasNext())
	{
		my $field = $iter->next();

		$fieldsFound += 1;

		if ($fieldsFound == 1)
		{
			$test->check("Unexpected field type 3", $field->getType() == $libgmsec_perl::Field::Type_I32);
			$test->check("Unexpected field name 3", $field->getName() eq "FIELD-3");
			$test->check("Unexpected field value 3", $field->getIntegerValue() == 3);
		}
		else
		{
            libgmsec_perl::logInfo("Unexpected field: " . $field->getName());
			$test->check("Unexpected field", 0);
		}
	}

	$test->check("Unexpected field count", $fieldsFound == 1);
}


sub test_reset
{
	libgmsec_perl::logInfo("Test reset()");

	my ($test) = @_;

	my $msg = libgmsec_perl::Message->new();
	$msg->addField( libgmsec_perl::StringField->new("FIELD-1", "1") );
	$msg->addField( libgmsec_perl::StringField->new("FIELD-2", "2") );
	$msg->addField( libgmsec_perl::I32Field->new("FIELD-3", 3) );

	my $iter = $msg->getFieldIterator( $libgmsec_perl::MessageFieldIterator::Selector_ALL_FIELDS );

	my $fieldsFound = 0;

	while ($iter->hasNext())
	{
		$iter->next();
		$fieldsFound += 1;
	}

	$test->check("Unexpected field count", $fieldsFound == 3);

	$test->check("Should no more field(s) to iterate over", $iter->hasNext() == 0);

	$iter->reset();

	$test->check("Should have field(s) to iterator over", $iter->hasNext());

	$fieldsFound = 0;

	while ($iter->hasNext())
	{
		$iter->next();
		$fieldsFound += 1;
	}

	$test->check("Unexpected field count", $fieldsFound == 3);
}


Test::driver(\&Test_MessageFieldIterator);
